### DISCUSSION TITLE
can it be true that use IISModule in kestrel?

### DISCUSSION THREAD

**Discussion by ljzj2** [dotnet/aspnetcore#56869] [Q&A]
in IIS I can use IISModule which is written in c++.
and in keystrel can it use c++ too?

what I want is to:
1.use c++ dll. so I can protect my code not to be changed.
2.do with some url forexample :if the url is `/recg` then I will use c++ to do it. and write response .
3.in linux I can use it too(with dll in linux)

**Comment by gfoidl** [comment]
Although I'm not entirely clear what you're asking for I'll try to answer.

>IISModule in kestrel

Both are implementations of a "server" in ASP.NET Core and can't be stacked together.
so you can't use IISModule in Kestrel.

> IIS I can use IISModule which is written in c++

The native part of the IISModule (that is written in C++) is needed to bridge over to the managed server part in ASP.NET Core (that is written in C#).

> use c++ dll. so I can protect my code not to be changed

What do you mean by "not to be changed"? 
I guess you mean something like "obfuscation" -- I'm not repeating the countless arguments why that doesn't give real guarantees for code inspection, change, etc. (just use your favorite search engine to find the topics).

In .NET you can use [ASP.NET Core support for Native AOT](https://learn.microsoft.com/en-us/aspnet/core/fundamentals/native-aot?view=aspnetcore-8.0) to get native binaries. In regards your concern they are equal to code written in C++.

Besides that server code usually runs on a server (😉), so you or the admin should be under control who has access to the binaries.
So that's a better mechanism to protect your intellectual property than w/ obfuscation, native code generation, etc.

> do with some url forexample :if the url is `/recg` then I will use c++ to do it. and write response .

To have NativeAOT only for certain urls isn't supported directly.
But can create a native DLL (e.g. written in C++) that holds the logic, and that is called from the managed request pipeline (but I doubt it is worth the trouble, see above for obfuscation).

> in linux I can use it too(with dll in linux)

NativeAOT can be compiled for Linux too.
Managed code runs as is (when the runtime is installed), BTW.

**Reply by martincostello** [reply]
Depending on what the OP is actually getting at, another answer separate from native AoT could be to write a custom middleware that delegates the "secret" implementation to a native library written with C++ with P/Invoke.

**Reply by ljzj2** [reply]
@martincostello  thank you for your answer
 how to write delegate? 

in c++
`std::string getResponse(std::string url,const unsigned char* req);`

I can only use this right? I can not use the `context` (from `app.Use((context,next)`)as parameter to the c++ code.  
there is `IHttpRequest` in `httpserv.h` 
I can only in 
```
app.Use((context,next)=>{
var url=context.Request.Path;
if(url=="/validate"){
       //getResponse
}
});
```
right?

**Reply by martincostello** [reply]
The documentation can be found here: [Platform Invoke (P/Invoke) - learn.microsoft.com](https://learn.microsoft.com/dotnet/standard/native-interop/pinvoke)

**Reply by ljzj2** [reply]
@gfoidl  can `IHttpContext` in asp.netcore be parameter to c++ func?

**Reply by gfoidl** [reply]
Not directly, as `IHttpContext` is not blitable. You have to define a custom struct and pass this to the C++ function (which itstelf should be `extern "C"`).

But is all that worth the trouble? See other comments from me and Martin.

**Reply by gfoidl** [reply]
See https://github.com/dotnet/aspnetcore/discussions/56886

**Comment by ljzj2** [comment]
thank you for your answer.
you suggest me to use Native AOT and what I want is to protect the part of the code. If one web has razor pages and controllers . I want to protect if the url is `/myvalidate` If user get this url then I want to response something.

I will set this repo(my project) in github for public but I do not users who use it change the response html of the url `/myvalidate`

**Reply by martincostello** [reply]
I don't think that's realistically 100% possible to defend against if the user is hosting your code themselves - a user could always add a custom middleware that buffers the response, and then run code after your response is written to the buffer and change it.

**Reply by ljzj2** [reply]
@martincostello  can `IHttpContext` be parameter to c++ func?

**Reply by gfoidl** [reply]
https://github.com/dotnet/aspnetcore/discussions/56886
